### PR TITLE
[aotautograd] Don't recompute reads of mutated inputs or mutable buffers

### PR DESCRIPTION
**Pull request by qchip**
Summary:
The AOT min-cut partitioner may recompute a value in the backward rather than save
it. That is equivalent only when re-execution is value-preserving, which fails for a
read of a tensor mutated in between: recompute defers the read past the mutation, so
the backward observes post-update contents. The symptom is silently wrong gradients,
or a version-counter error when the input reaches the backward through
`SavedVariable`. Two cases arise. Inputs the forward mutates are reported precisely by
`input_info.mutates_data`. Module buffers mutated by another compiled region or by
eager code are invisible in this graph, so they are handled conservatively behind
`ban_recompute_through_buffers`, off by default. The fix pins such reads `MUST_SAVE`,
following aliasing ops through to the op that materializes data, since pinning a view
pins an alias rather than a snapshot. A `TestPartitioning` regression test covers both
cases and the view path.

## Details

Pinning readers rather than the input suffices: every backward chain reaching the
input passes through one, so the input need not stay live.

The functionalized in-place chain is exempt, but only the nodes the epilogue alone
consumes: a value written back to the input can equally feed the forward's own
computation. The walk also follows only nodes that themselves take a mutated input,
not the mutation value's whole ancestor cone.

`memory_budget == 0` early-returned before `solve_min_cut`, bypassing the tag; it now
retains `MUST_SAVE` forward nodes. A tuple-producing op cannot be saved as a tensor,
so pins land on its `getitem` leaves.

Dynamo passes params and buffers as ordinary graph inputs under
`inline_inbuilt_nn_modules`, so buffers are recovered as the static inputs that are
not `nn.Parameter`. Effect tokens are prepended, so positions resolve to node names in
`graph_compile`.

Parameters stay exempt; `_apply_optimizer_in_backward` breaks that assumption and is a
pre-existing hazard not addressed here.

Test Plan:
```
buck2 test @//mode/opt fbcode//caffe2/test/functorch:test_aotdispatch \
    -- --regex 'no_recompute_through_mutated_buffer'
```

Each case verified non-vacuous by disabling its guard. Partitioner and AC suites pass.

Differential Revision: D118411035

### CODE REVIEW

**Review by ezyang — COMMENTED** [review]
Review by Astra agent.

Found two remaining correctness gaps in the mutation protection and a compilation failure in the budget-zero path; details are inline. Reproduced on a locally patched PyTorch 2.14 CPU wheel with the PR partitioner source (omitting references to the unavailable, unrelated distributed wait_tensors op). The PR's added regression test passes in that environment; this was not a full source-build or test-suite run.

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -2417,6 +2418,79 @@ def force_save_bw_mutation_src(joint_module: fx.GraphModule) -> None:
...
+        mutation_chain.add(node)
+        stack.extend(node.all_input_nodes)
+
+    for node in joint_module.graph.nodes:
+        if node.op != "call_function" or node in mutation_chain:
+            continue
+        if _has_tag_must_be_in_backward(node):
+            continue
+        if any(inp in mutated_inputs for inp in node.all_input_nodes):
+            node.meta["recompute"] = CheckpointPolicy.MUST_SAVE
```
Comment (ezyang) [inline torch/_functorch/partitioners.py]: [P1] Follow aliases through to reads that materialize independent values

Marking only the direct reader does not preserve the input value when that reader is a view. For a zero-initialized buffer, `cond = counts.view(16) > .5; y = torch.where(cond, x * .5, x).sum(); counts.add_(1); return y` still produces gradients of **0.5 instead of eager's 1.0** at memory budgets 0 and 0.05. At budget zero the saved view aliases the mutated buffer, so recomputing the predicate observes the updated contents. Moving the mutation outside the compiled forward also still raises the version-counter error with `ban_recompute_through_buffers=True`. Please propagate the protection through aliasing operations to the actual data reads, or otherwise save an independent snapshot.

— Astra agent

File: torch/_functorch/partitioners.py [hunk 2]
```diff
@@ -2417,6 +2418,79 @@ def force_save_bw_mutation_src(joint_module: fx.GraphModule) -> None:
...
+        node = stack.pop()
+        if node in mutation_chain:
+            continue
+        if not any(inp in mutated_inputs for inp in node.all_input_nodes):
+            continue
+        mutation_chain.add(node)
+        stack.extend(node.all_input_nodes)
+
+    for node in joint_module.graph.nodes:
+        if node.op != "call_function" or node in mutation_chain:
```
Comment (ezyang) [inline torch/_functorch/partitioners.py]: [P1] Do not exempt mutation values that also feed backward computation

Being the source of the mutation epilogue does not mean this value is used only for the update. Starting from a zero buffer, `counts.add_(1); return torch.where(counts > 1.5, x * .5, x).sum()` functionalizes to an `add` feeding both the predicate and the final `copy_`. This exemption leaves the `add` recomputable, so backward adds 1 to the already-updated buffer and evaluates a different predicate. I reproduced gradients of **0.5 instead of eager's 1.0** at budgets 0 and 0.05 with `keep_inference_input_mutations=True`. Please restrict the exemption to values that cannot participate in backward recomputation, or protect the reads through this chain too.

— Astra agent

File: torch/_functorch/partitioners.py [hunk 3]
```diff
@@ -3520,7 +3594,15 @@ def choose_saved_values_set(
...
         )
     if memory_budget == 0:
-        return node_info.inputs
+        # Saving only the inputs means recomputing everything else, but MUST_SAVE
+        # marks values whose recomputation is not value-preserving -- a read of an
+        # input the forward mutates, for one. Those stay saved at any budget.
+        return node_info.inputs + [
+            n
+            for n in joint_graph.nodes
+            if n.meta.get("recompute") == CheckpointPolicy.MUST_SAVE
```
Comment (ezyang) [inline torch/_functorch/partitioners.py:3646]: [P2] Save tensor outputs rather than tuple-valued MUST_SAVE nodes

This list can contain tuple-producing operations newly marked by `force_save_fw_mutated_input_readers`. For example, `cond = counts.sort().values > .5; y = torch.where(cond, x * .5, x).sum(); counts.add_(1); return y` marks `aten.sort.default` MUST_SAVE. With `activation_memory_budget=0`, compilation then fails in `_categorize_saved_tensors_for_backward` with `expected one saved_tensor_is_graph_input entry per saved tensor, got 0 != 1`, because the saved value is the sort tuple rather than a tensor. The same example compiles and matches eager at budget 0.05. Please expand multi-output nodes to their tensor leaves before constructing the budget-zero saved-value set.

— Astra agent

**Review by ezyang — CHANGES_REQUESTED** [review]
Human here: this patch is very very long, let's make a frontier LLM happy first, but when I do human review I'll be trying to understand why we need such a long PR

**Comment by qchip** [comment]
> Human here: this patch is very very long, let's make a frontier LLM happy first, but when I do human review I'll be trying to understand why we need such a long PR

Thanks! Made the PR concise, also addressed the other 3 review comments.

File: torch/_functorch/partitioners.py (hunk 1 above)
Comment (qchip) [reply torch/_functorch/partitioners.py]: fixed

File: torch/_functorch/partitioners.py (hunk 2 above)
Comment (qchip) [reply torch/_functorch/partitioners.py]: fixed

File: torch/_functorch/partitioners.py (hunk 3 above)
Comment (qchip) [reply torch/_functorch/partitioners.py:3646]: fixed

File: torch/_functorch/_aot_autograd/graph_compile.py [hunk 4]
```diff
@@ -1984,6 +1984,33 @@ def _partition_joint_graph_into_fw_bw(
...
+    # the partition_fn signature, which custom partitioners also implement.
+    # Tokens for effectful ops are prepended to the graph inputs, so a position in
+    # input_info is num_tokens placeholders further along. Resolve to node names
+    # here, where that offset is known, rather than making the partitioner reason
+    # about which index space it has been handed.
+    _placeholders = fx_g.graph.find_nodes(op="placeholder")
+
+    def _input_names(indices: Iterable[int]) -> list[str]:
+        return [
+            _placeholders[num_tokens + i].name
```
Comment (ezyang) [inline torch/_functorch/_aot_autograd/graph_compile.py:2000]: *(Posted by Opus 5.5, an AI assistant, on behalf of @ezyang.)*

**[P1] Outer `input_info` positions don't line up with the joint graph's placeholders when there are tensor-subclass inputs**

`fw_metadata.input_info` is indexed by outer (subclass) inputs, but `fx_g`'s placeholders have subclasses split into their inner tensors. `mutated_inp_runtime_indices` a few lines up handles this with `compute_inner_mutated_inp_indices_from_subclass_meta(fw_metadata, inner_meta)`, and this code should do the same, e.g. use `inner_meta.input_info`.

Repro: `f(x, counts)` where `x = TwoTensor(ones(4), ones(4)).requires_grad_()`, and the body is `xh = x * .5; cond = counts[idx] > .5; y = where(cond, xh, x).sum(); counts.add_(1)`, run at `activation_memory_budget=0`. The placeholders are `[x.a, x.b, idx, counts]`, but `aot_mutated_input_names == ['primals_2']`, which is `x.b` rather than `counts`.
- With `keep_inference_input_mutations=True`, the `copy_` scan happens to find the right placeholder, which hides the bug.
- With `keep_inference_input_mutations=False`, the backward still contains `aten.index` and fails with "modified by an inplace operation", i.e. the exact bug this PR fixes.

`aot_config.buffer_input_indices` has the same problem. It is in the original flat-arg index space plus `num_tokens` and is never remapped for subclasses, whereas `static_input_indices` is remapped in `graph_capture_wrappers.py`.

Also, please assert rather than silently dropping out-of-range positions (`if 0 <= num_tokens + i < len(_placeholders)`). This bug would have shown up immediately.

File: torch/_functorch/partitioners.py [hunk 5]
```diff
@@ -2417,6 +2418,122 @@ def force_save_bw_mutation_src(joint_module: fx.GraphModule) -> None:
...
+    # the backward still observes the post-mutation contents. Walk through
+    # aliasing ops to the first reads that produce independent data, and pin
+    # those.
+    tainted = OrderedSet(mutated_inputs)
+    stack = list(mutated_inputs)
+    while stack:
+        for user in stack.pop().users:
+            if user.op != "call_function" or user in mutation_chain:
+                continue
+            if _returns_alias(user):
```
Comment (ezyang) [inline torch/_functorch/partitioners.py:2528]: *(Posted by Opus 5.5, an AI assistant, on behalf of @ezyang.)*

**[P1] Reads through `chunk`/`split`/`unbind` still see the mutated storage**

`_returns_alias` relies on the op schema. `split.Tensor` returns `Tensor(a)[]`, so the walk marks it as a view, but its users are `operator.getitem`, which has no `_schema`. So the `getitem` counts as a node that materializes data and is marked `MUST_SAVE`, even though it is a view of the mutated buffer.

Repro: `cond = self.counts.chunk(2)[0] > 0.5; y = torch.where(cond, x * 0.5, x).sum(); self.counts.add_(1)`, with a zero-initialized float64 buffer and `keep_inference_input_mutations=True`. At `activation_memory_budget=0` the gradient is **0.5 instead of eager's 1.0**, the same as on trunk. Budgets 0.05 and 1.0 are fine.

At minimum, carry the taint through a `getitem` whose input is tainted. A sturdier option is to decide aliasing by storage (`get_node_storage` on the fake values) instead of by schema, since that covers every view-producing op the same way.

File: torch/_functorch/partitioners.py [hunk 6]
```diff
@@ -2417,6 +2418,122 @@ def force_save_bw_mutation_src(joint_module: fx.GraphModule) -> None:
...
+        if isinstance(dst, fx.Node) and dst.op == "placeholder":
+            mutated_inputs.add(dst)
+            epilogue_copies.add(node)
+            if isinstance(src, fx.Node):
+                mutation_values.add(src)
+
+    if not mutated_inputs:
+        return
+
+    # Exempt the functionalized in-place chain itself: those nodes read the
```
Comment (ezyang) [inline torch/_functorch/partitioners.py:2484]: *(Posted by Opus 5.5, an AI assistant, on behalf of @ezyang.)*

**[simplification] The mutation-chain exemption and the escaping fixpoint have no effect and can be deleted**

After the escaping filter, every node left in `mutation_chain` is used only by other chain nodes or by the epilogue `copy_`, so the backward never needs it. Marking such a node `MUST_SAVE` costs nothing: min-cut only saves nodes the backward uses, and the budget-0 list is pruned of saved values the backward doesn't use.

I checked this by replacing this block with `mutation_chain = OrderedSet()`. The PR's regression test still passes, and the saved forward outputs are identical to this PR's for a buffer EMA update (`ema.add_(1).mul_(0.9)`) at budgets 0, 0.05 and 1.0, and for a BatchNorm ResNet at budgets 1.0 and 0.5.

The walk also stops at the first node that doesn't read a mutated input directly, so for multi-op updates the chain was almost empty anyway.

File: torch/_functorch/partitioners.py [hunk 7]
```diff
@@ -2417,6 +2418,122 @@ def force_save_bw_mutation_src(joint_module: fx.GraphModule) -> None:
...
+    # sound option is to treat every buffer as mutable.
+    if config.ban_recompute_through_buffers:
+        unsafe_names.update(joint_module.meta.get("aot_buffer_input_names") or [])
+
+    if unsafe_names:
+        for node in joint_module.graph.find_nodes(op="placeholder"):
+            if node.name in unsafe_names:
+                mutated_inputs.add(node)
+
+    for node in reversed(joint_module.graph.nodes):
```
Comment (ezyang) [inline torch/_functorch/partitioners.py:2465]: *(Posted by Opus 5.5, an AI assistant, on behalf of @ezyang.)*

**[simplification] The `copy_` epilogue scan is redundant with `input_info`**

Once `mutation_chain` is gone (see the comment on the exemption block), this loop's only job is to discover mutated inputs, which `aot_mutated_input_names` already reports, as the comment above says. With this loop also removed, the PR's regression test still passes. Right now it is effectively a second source of truth, and it is what hid the subclass indexing bug in the in-graph case.

With both removals, `force_save_fw_mutated_input_readers` is: collect the unsafe placeholders, walk through views, and pin the first node that materializes data. That's about 28 non-comment lines.

File: torch/_functorch/aot_autograd.py [hunk 8]
```diff
@@ -1007,13 +1007,29 @@ def prepare_aot_config(
...
             break
 
+    # Which flat inputs are module buffers. Under inline_inbuilt_nn_modules Dynamo
+    # hands params and buffers over as ordinary graph inputs, so buffers_len above
+    # is 0 and the split has to be recovered from the values: the module attributes
+    # are exactly the static inputs, and the ones that are not Parameters are
+    # buffers. Buffers carry mutable state that another region -- or eager code --
+    # can update between a forward and its backward, so the partitioner must not
+    # recompute reads of them.
+    buffer_input_indices = [
```
Comment (ezyang) [inline torch/_functorch/aot_autograd.py:1017]: *(Posted by Opus 5.5, an AI assistant, on behalf of @ezyang.)*

**[scope] Consider moving the `ban_recompute_through_buffers` plumbing to a follow-up PR**

It is off by default, but it touches `aot_autograd.py`, `schemas.py` (`AOTConfig` and `CacheableAOTConfig`) and `config.py`, and it needs the same subclass remapping fix as `input_info` (see the comment in `graph_compile.py`). The natural place to compute it is next to `static_input_indices` in `ViewAndMutationMeta`, where the remapping for effect tokens and subclasses already happens. Then this PR could be just the default-on fix for inputs mutated in this graph.

File: test/functorch/test_aotdispatch.py [hunk 9]
```diff
@@ -7645,6 +7645,290 @@ def bw_compiler(gm, _):
         # so the producer op does not appear in the backward graph.
         self.assertNotIn("topk", bw_graph["gm"].code)
 
+    @unittest.skipIf(not USE_NETWORKX, "networkx not available")
+    def test_min_cut_partitioner_no_recompute_through_mutated_buffer(self):
```
Comment (ezyang) [inline test/functorch/test_aotdispatch.py:7649]: *(Posted by Opus 5.5, an AI assistant, on behalf of @ezyang.)*

**[test] Please split this into separate tests that share one helper**

This single method is about 285 lines covering six scenarios, and it redefines the same `aot_module_simplified` backend closure three times. One parametrized helper, "compile with min-cut at budget B and compare grads with eager", plus one short test per scenario (in-graph mutation, out-of-graph mutation, view read, shared update, tuple read, trunk behaviour) would be much shorter and would make failures easier to pin down. It would also be a good place to add the subclass and `chunk` cases from the other comments.

**Review by ezyang — CHANGES_REQUESTED** [review]
*(Posted by Opus 5.5, an AI assistant, on behalf of @ezyang.)*

Thanks for the fixes to the earlier round. The approach is sound: on trunk the PR's regression test fails (the backward reads the buffer again) and with this PR it passes, and a small BatchNorm ResNet saves the same tensors and bytes as trunk at budgets 1.0 and 0.5. But two correctness holes remain, and a good chunk of the new code turns out to be unnecessary, which is most of the answer to "why is this so long". Details are inline.

Summary:
1. **Subclass inputs**: the `input_info` positions are outer positions, but they are used to index the joint graph's placeholders, where subclasses are split into their inner tensors. So the wrong placeholder gets protected. `buffer_input_indices` has the same problem.
2. **Reads through `chunk`/`split`/`unbind`** are still recomputed from an alias of the mutated storage at budget 0.
3. **Unnecessary code**: the mutation-chain exemption plus the escaping fixpoint, and the `copy_` epilogue scan, can all be deleted with no change in behaviour. That shrinks `force_save_fw_mutated_input_readers` to about 28 non-comment lines.
4. **Scope**: I'd split the opt-in `ban_recompute_through_buffers` plumbing (`schemas.py`, `aot_autograd.py`, `config.py`, `CacheableAOTConfig`) into a follow-up PR, so this PR is just the default-on fix for inputs mutated in this graph.
5. **Test**: about 285 lines in one method, with the backend closure copied three times. Please split it into separate tests that share one helper.

Minor: the pin overwrites a user's `MUST_RECOMPUTE` without warning, and the budget-0 change now also keeps user checkpoint-policy `MUST_SAVE` nodes. That's probably right, but it changes behaviour and the PR description should mention it.

How I tested: I applied this PR's Python changes to a torch 2.14 CPU wheel and ran the extracted regression test plus small probes, comparing against unpatched code. This was not a source build or a full test-suite run.

**Comment by ezyang** [comment]
next review round will be the last, I will unsubscribe from this PR if opus 5.5 doesn't give a clean bill of health
